In [1]:
!pip install -q pyomo

     |████████████████████████████████| 9.7 MB 21.1 MB/s 
     |████████████████████████████████| 49 kB 6.0 MB/s 


In [2]:
import pandas as pd


In [3]:
from pyomo.environ import *

In [4]:
!wget -N -q "https://ampl.com/dl/open/ipopt/ipopt-linux64.zip"
!unzip -o -q ipopt-linux64

In [5]:
def pyomo1():
  global model
  model=ConcreteModel()
  model.x=Var(range(5),initialize=2)
  model.x[0].setub(5)
  model.x[0].setlb(-5)
  model.x[2].setlb(0)
  model.x[2].setub(3)
  def obj(model):
    return (model.x[0]-1)**2 + (model.x[0]-model.x[1])**2 + (model.x[1]-model.x[2])**2 + (model.x[2]-model.x[3])**4 + (model.x[3]-model.x[4])**4
  model.objective=Objective(rule=obj,sense=minimize)
  model.constr1=Constraint(expr=model.x[0]+model.x[1]**2+model.x[2]**3 - 2 -3*(2**0.5)==0)
  model.constr2=Constraint(expr=model.x[1]-model.x[2]**2+model.x[3]+2-2*(2**0.5)==0)
  model.constr3=Constraint(expr=model.x[0]*model.x[4]-2==0)

In [6]:
def result1():
  solver = SolverFactory('ipopt', executable='/content/ipopt')
  global results
  results = solver.solve(model)
  print(results)
  print('\n objective function value = ', model.objective())
  print('\n optimal values of the optimization variables')
  for i in range(5):
    print('x',i+1,'  =   ',model.x[i].value)
  print('\n constraints')
  print(model.constr1.display())
  print(model.constr2.display())
  print(model.constr3.display())

In [7]:
  pyomo1()
  result1()


Problem: 
- Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 5
  Sense: unknown
Solver: 
- Status: ok
  Message: Ipopt 3.12.13\x3a Optimal Solution Found
  Termination condition: optimal
  Id: 0
  Error rc: 0
  Time: 0.033014535903930664
Solution: 
- number of solutions: 0
  number of solutions displayed: 0


 objective function value =  0.07877682087105699

 optimal values of the optimization variables
x 1   =    1.1911274562403504
x 2   =    1.3626031649434767
x 3   =    1.4728179315306016
x 4   =    1.6350166192407931
x 5   =    1.6790814362660684

 constraints
constr1 : Size=1
    Key  : Lower : Body                   : Upper
    None :   0.0 : 1.7763568394002505e-15 :   0.0
None
constr2 : Size=1
    Key  : Lower : Body                   : Upper
    None :   0.0 : -4.440892098500626e-16 :   0.0
None
constr3 : Size=1
    Key  : Lower : Body                   : Upper
    None :   0.0 : -3.774758283725532e-15 :   0.0
Non

$\huge{solution \ \ of \ \ 1.3}$

solver Status: ok

Time taken to solve: 0.02970123291015625


solver Termination condition: optimal



$\huge{solution \ \ of \ \ 1.4}$

 objective function value =  0.07877682087105699

 optimal values of the optimization variables


x 1   =    1.1911274562403504

x 2   =    1.3626031649434767

x 3   =    1.4728179315306016

x 4   =    1.6350166192407931

x 5   =    1.6790814362660684

$\huge{solution \ \ of \ \ 1.5}$

all constrained are active 
 since we get approx same value of constrained as mention in question i.e 0 

 for constraints 1 we get 1.7763568394002505e-15 ≈0

 for constraints 2 we get -4.440892098500626e-16 ≈0

 for constraints 3 we get -3.774758283725532e-15 ≈0

In [8]:
def pyomo(p):
  global model
  model=ConcreteModel()
  model.x=Var(range(5),initialize=2**p)
  model.x[0].setub(5)
  model.x[0].setlb(-5)
  model.x[2].setlb(0)
  model.x[2].setub(3)
  def obj(model):
    return (model.x[0]-1)**2 + (model.x[0]-model.x[1])**2 + (model.x[1]-model.x[2])**2 + (model.x[2]-model.x[3])**4 + (model.x[3]-model.x[4])**4
  model.objective=Objective(rule=obj,sense=minimize)
  model.constr1=Constraint(expr=model.x[0]+model.x[1]**2+model.x[2]**3 - 2 -3*(2**0.5)==0)
  model.constr2=Constraint(expr=model.x[1]-model.x[2]**2+model.x[3]+2-2*(2**0.5)==0)
  model.constr3=Constraint(expr=model.x[0]*model.x[4]-2==0)

In [9]:
def result():
  solver = SolverFactory('ipopt', executable='/content/ipopt')
  global results
  results = solver.solve(model)
  print(results)
  print('\n objective function value = ', model.objective())
  print('\n optimal values of the optimization variables')
  for i in range(5):
    print('x',i+1,'  =   ',model.x[i].value)
  print('\n constraints')
  print(model.constr1.display())
  print(model.constr2.display())
  print(model.constr3.display())

In [10]:
starting_point=[1,2,4,8,10,12]
optimal_solution=[]
termination_condition=[]
varx=[[] for i in range(5)]
time=[]
status=[]
starting2=[[2**i for j in range(5)] for i in starting_point]

In [11]:
for i in starting_point:
  pyomo(i)
  print(' \n \n \n \n for starting point=2**',i)
  result()
  optimal_solution.append(model.objective())
  termination_condition.append(results.solver.termination_condition)
  time.append(results.solver.time)
  status.append(results.solver.status)
  for j in range(5):
    varx[j].append(model.x[j].value)

 
 
 
 
 for starting point=2** 1

Problem: 
- Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 5
  Sense: unknown
Solver: 
- Status: ok
  Message: Ipopt 3.12.13\x3a Optimal Solution Found
  Termination condition: optimal
  Id: 0
  Error rc: 0
  Time: 0.029161453247070312
Solution: 
- number of solutions: 0
  number of solutions displayed: 0


 objective function value =  0.07877682087105699

 optimal values of the optimization variables
x 1   =    1.1911274562403504
x 2   =    1.3626031649434767
x 3   =    1.4728179315306016
x 4   =    1.6350166192407931
x 5   =    1.6790814362660684

 constraints
constr1 : Size=1
    Key  : Lower : Body                   : Upper
    None :   0.0 : 1.7763568394002505e-15 :   0.0
None
constr2 : Size=1
    Key  : Lower : Body                   : Upper
    None :   0.0 : -4.440892098500626e-16 :   0.0
None
constr3 : Size=1
    Key  : Lower : Body                   : Upper
    None :   0.0 : 

In [12]:
data={'optimal solution':optimal_solution,'var x1':varx[0],'var x2':varx[1],'var x3':varx[2],'var x4':varx[3],'var x5':varx[4],'time taken to solve':time,'status':status,'termination condition':termination_condition,'starting point':starting2}

In [13]:
df=pd.DataFrame.from_dict(data).set_index('starting point')

In [14]:
df

optimal solution    var x1    var x2  \
starting point                                                         
[2, 2, 2, 2, 2]                         0.078777  1.191127  1.362603   
[4, 4, 4, 4, 4]                         0.078777  1.191127  1.362603   
[16, 16, 16, 16, 16]                    0.078777  1.191127  1.362603   
[256, 256, 256, 256, 256]               0.078777  1.191127  1.362603   
[1024, 1024, 1024, 1024, 1024]          0.078777  1.191127  1.362603   
[4096, 4096, 4096, 4096, 4096]          0.078777  1.191127  1.362603   

                                  var x3    var x4    var x5  \
starting point                                                 
[2, 2, 2, 2, 2]                 1.472818  1.635017  1.679081   
[4, 4, 4, 4, 4]                 1.472818  1.635017  1.679081   
[16, 16, 16, 16, 16]            1.472818  1.635017  1.679081   
[256, 256, 256, 256, 256]       1.472818  1.635017  1.679081   
[1024, 1024, 1024, 1024, 1024]  1.472818  1.635017  1.679081   
[4096, 4096, 4096, 4096, 4096]  1.472818  1.635017  1.679081   

                                time taken to solve status  \
starting point                                               
[2, 2, 2, 2, 2]                            0.029161     ok   
[4, 4, 4, 4, 4]                            0.042477     ok   
[16, 16, 16, 16, 16]                       0.044585     ok   
[256, 256, 256, 256, 256]                  0.058695     ok   
[1024, 1024, 1024, 1024, 1024]             0.079415     ok   
[4096, 4096, 4096, 4096, 4096]             0.044017     ok   

                               termination condition  
starting point                                        
[2, 2, 2, 2, 2]                              optimal  
[4, 4, 4, 4, 4]                              optimal  
[16, 16, 16, 16, 16]                         optimal  
[256, 256, 256, 256, 256]                    optimal  
[1024, 1024, 1024, 1024, 1024]               optimal  
[4096, 4096, 4096, 4096, 4096]               optimal

$\huge{solution \ \ of \ \ 1.6}$





|starting point|optimal solution|var x1|var x2|var x3|var x4|var x5|time taken to solve|status|termination condition|
|---|---|---|---|---|---|---|---|---|---|
|2,2,2,2,2|0\.07877682087105699|1\.1911274562403504|1\.3626031649434767|1\.4728179315306016|1\.6350166192407931|1\.6790814362660684|0\.0375521183013916|ok|optimal|
|4,4,4,4,4|0\.07877682087105696|1\.1911274562403502|1\.3626031649434764|1\.4728179315306016|1\.6350166192407936|1\.6790814362660686|0\.0430905818939209|ok|optimal|
|16,16,16,16,16|0\.07877682087105699|1\.1911274562403504|1\.3626031649434767|1\.4728179315306016|1\.6350166192407931|1\.6790814362660684|0\.05226635932922363|ok|optimal|
|256,256,256,256,256|0\.07877682087142031|1\.1911274559458875|1\.3626031648627808|1\.4728179316111962|1\.635016619557219|1\.6790814366787805|0\.07756876945495605|ok|optimal|
|1024,1024,1024,1024,1024|0\.0787768208710788|1\.1911274548626596|1\.3626031645866354|1\.4728179318918395|1\.6350166206615917|1\.6790814382064987|0\.1033334732055664|ok|optimal|
|4096,4096,4096,4096,4096|0\.07877682153434043|1\.1911274444329922|1\.362603152254644|1\.4728179414916138|1\.635016658269562|1\.6790814465461419|0\.07150077819824219|ok|optimal|





here in first initial value(2,2,2,2,2)  it is nearer to optimal solution so it takes less time to solve if we take initial value far away from optimal point it takes more time to solve our problem graph is simple so it not takes more time in each initial value but if graph is not simple then it takes too much time in our problem time difference is not so much differ by taking differnt initial value we can see that from above table if initial value is (2,2,2,2,2) then it takes less time then other